### PR TITLE
Attempt to update for SE-0103 (@noescape by default)

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test linux

**Comment by slavapestov** [comment]
@gparker42 Here's the second fix.

**Comment by slavapestov** [comment]
@swift-ci Please test linux

**Comment by slavapestov** [comment]
@swift-ci Please test linux

**Comment by slavapestov** [comment]
@gonzalolarralde Do you mind reviewing this? It builds upon your previous patch by properly marking all functions that previously were not @noescape as @escaping. Otherwise, almost everything becomes non-escaping, since that's the default.

**Comment by dgrove-oss** [comment]
does this need to go to the Swift3 branch as well?

File: src/swift/Queue.swift [hunk 1]
```diff
@@ -271,8 +271,8 @@ public extension DispatchQueue {
 	private func _syncHelper<T>(
 		fn: (DispatchWorkItem) -> (), 
 		flags: DispatchWorkItemFlags,
-		execute work: @noescape () throws -> T,
-		rescue: ((Swift.Error) throws -> (T))) rethrows -> T
+		execute work: () throws -> T,
+		rescue: @escaping ((Swift.Error) throws -> (T))) rethrows -> T
```
Comment (gonzalolarralde) [inline src/swift/Queue.swift]: I think `rescue` could be `@noescape`. I can't see nothing keeping a reference to it.

**Comment by gonzalolarralde** [comment]
Hi @slavapestov, sorry my patch wasn't complete at that time. The changes in this PR makes total sense to me. 

I took a look on the rest of the project and I found the following declarations that were `@escaping` (default, actually) before `SE-0103` and that I think they should remain `@escaping`.

* [`dispatch_data_create(…, destructor)`](https://github.com/apple/swift-corelibs-libdispatch/blob/95875d7bed3f66ed8186f9f77bad8ad23f80d152/src/swift/Private.swift#L72) along with [`DispatchData.Deallocator.custom`](https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Data.swift#L36-L38)
* [`dispatch_queue_set_specific(…, destructor)`](https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Private.swift#L216) (there's no associated typealias / parameter for this one, it's just being called with [`Queue._destructDispatchSpecificValue`](https://github.com/apple/swift-corelibs-libdispatch/blob/1a7ff3f3e1073eb3352a56ab121ccfa712c42cef/src/swift/Queue.swift#L368-L372) as parameter)

And finally,

* [`dispatch_source_set_event_handler(…, handler)`](https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Private.swift#L270)
* [`dispatch_source_set_cancel_handler(…, handler)`](https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Private.swift#L276)
* [`dispatch_source_set_registration_handler(…, handler)`](https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Private.swift#L324)

along with

* [`DispatchSourceHandler`](https://github.com/apple/swift-corelibs-libdispatch/blob/55261225184e49c6a42c38bbedb144c2610def4a/src/swift/Wrapper.swift#L209)
* [`_DispatchBlock`](https://github.com/apple/swift-corelibs-libdispatch/blob/1a7ff3f3e1073eb3352a56ab121ccfa712c42cef/src/swift/Block.swift#L103)

I can create a new PR with this changes if you want. Just let me know.

Thanks!

**Pull request merged**